### ISSUE TITLE
Cross-compiling OpenDDS for the powerpc-604-linux-gnu platform fails during TAO IDL Compilation

### ISSUE THREAD

**Issue by PaulA-V** [objectcomputing/OpenDDS#2349]
Hello,

Here is the problem report form data pertaining to the issue I am experiencing.

This is OpenDDS version 3.15, released Dec 22 2020

    OpenDDS VERSION: 3.15

    TAO VERSION: 2.2a_p19

    HOST MACHINE and OPERATING SYSTEM: Linux 4.4.180-94.100-default #1 SMP x86_64 GNU/Linux

    TARGET MACHINE and OPERATING SYSTEM, if different from HOST: Linux ueipac 3.4.6-ueipac834x #12 PREEMPT Fri Jun 9 06:09:27 EDT 2017 ppc GNU/Linux

    COMPILER NAME AND VERSION (AND PATCHLEVEL): powerpc-604-linux-gnu-g++ (crosstool-NG crosstool-ng-1.23.0) 6.3.0

    CONTENTS OF $ACE_ROOT/ace/config.h [if you use a link to a platform-specific file, simply state which one]: See attached build configuration (produced by show_build_config.pl) 

    CONTENTS OF $ACE_ROOT/include/makeinclude/platform_macros.GNU (unless this isn't used in this case, e.g., with Microsoft Visual C++): See attached build configuration (produced by show_build_config.pl)

    CONTENTS OF $ACE_ROOT/bin/MakeProjectCreator/config/default.features (used by MPC when you generate your own makefiles): See attached build configuration (produced by show_build_config.pl)

    LEVEL OF URGENCY (LOW, MEDIUM, or HIGH): High (spent two weeks trying to get this to work)

    AREA/CLASS/EXAMPLE AFFECTED: [What example failed?  What module failed to compile?] Cross-compilation of OpenDDS failed.  Compilation of the PortableServer IDL failed.

    DOES THE PROBLEM AFFECT:
        COMPILATION? Yes
            If so, what do your $ACE_ROOT/ace/config.h and $ACE_ROOT/include/makeinclude/platform_macros.GNU contain? See attached build configuration (produced by show_build_config.pl)

        LINKING? No
            On Unix systems, did you run make realclean first? Yes

        EXECUTION? Unknown

        OTHER (please specify)?  IDL compilation fails

[Please indicate whether DDS, your application, or both are affected.]  DDS

    SYNOPSIS:
Cross-compiling OpenDDS for the powerpc-604-linux-gnu platform fails during TAO IDL Compilation

    DESCRIPTION:
I am attempting to cross-compile OpenDDS for the powerpc-604-linux-gnu platform.  The compilation fails at the TAO IDL compilation stage.  The ace_gperf hash utility produces errors on the first IDL compilation then bails.  Please find attached the build configuration and the output from the build.  I’ve followed the cross-compile instructions verbatim.  Fully aware that this platform is a new one on OpenDDS I looked through instructions and scoreboards of a number of the supported cross-compile platforms (VxWorks, Android) for a clue to anything that might be missing in the “straight” Linux cross-compile to no avail.  I’ve also scoured the OpenDDS mailing lists for similar issues and didn’t find anything similar to my issue.  According to the vendor of this platform (United Electronic Industries) building third party software is fairly straight forward with very little in the way of special compiler configuration parameters and/or macros etc. 

    REPEAT BY: [What you did to get the error; include test program or session transcript if at all possible.  ]  See attached log of the output from the build

    SAMPLE FIX/WORKAROUND: [If available ]  None.
 

Please review the configuration and compilation logs for any errors in the way the compilation was configured and/or carried out.  Thank you.

**NOTE: @mitza-oci started helping with this issue via the sourceforge.net opendds-main mailing list.  I am submitting the issue here due to deprecation of the sourceforge mailing lists.  We've removed the conversation from sourceforge.net.  I am replicating the initial submission and responses to provide a full record of the steps taken so far to resolve the issue.**

Regards,

Paul
[OpenDDS-3.15_ppc-604-linux_build-log-2021-01-22_05-30-43.zip](https://github.com/objectcomputing/OpenDDS/files/5876242/OpenDDS-3.15_ppc-604-linux_build-log-2021-01-22_05-30-43.zip)

**Comment by PaulA-V** [comment]
On January 25, 2021 at 11:37 AM @mitza-oci wrote:

_Hi Paul,

Do you have a log of the error you're seeing?  The description says "please find... output from the build" but I didn't see that attached.

I suggest you follow essentially the same steps we have documented here: https://opendds.org/quickstart/GettingStartedPi.html
 although you are not using a Raspberry Pi, it's still a Linux cross-compile.

If you need dedicated commercial support please see https://objectcomputing.com/products/opendds/opendds-consulting-and-support

Thanks,
Adam Mitz
Principal Software Engineer and Partner
Object Computing, Inc._

**Comment by PaulA-V** [comment]
The logs requested are attached to the original post.

**Comment by PaulA-V** [comment]
On January 25, 2021 at 12:45 PM @mitza-oci wrote:

Hi Paul,

The log contains this:
ln -s ../../apps/gperf/src/Linux/ace_gperf ../../../bin/Linux/ace_gperf
So it's putting things in a "Linux" subdirectory instead of where they should be.  This could be due to the ARCH environment variable is set.  ACE's makefiles use that variable, try unsetting it.
But based on what we see later in the log, there is some file here: ACE_wrappers/build/host/bin/ace_gperf
Try running "file" on this to see what it is.


Thanks,
Adam Mitz
Principal Software Engineer and Partner
Object Computing, Inc.

**Comment by PaulA-V** [comment]
On January 25, 2021 at 1:59 PM @PaulA-V wrote:

Adam,

I did notice during the initial run of the build ace_gperf wasn't found.  I noticed both it and tao_idl were in the Linux subdirectory.  I added the symbolic link to ace_gperf which is what you are seeing in the log.  I'll give your suggestion a try and let you know how it goes.  Thanks for the help!


Regards,

Paul

**Comment by PaulA-V** [comment]
On January 25, 2021 at 5:29 PM @PaulA-V wrote:

Hi Adam,

Un-setting ARCH eliminated the gperf errors.  However, the FowardRequest.pidl compilation still fails.  There's a little more information about the failure this time.  The logs are in the attached file.  I've also included the output from the configure script this time (config_log.txt).

GNUmakefile: /projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/target/TAO/tao/PortableServer/GNUmakefile.PortableServer_Idl MAKEFLAGS=w

/projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/host/bin/tao_idl -Wb,pre_include=ace/pre.h -Wb,post_include=ace/post.h -I../.. -g /projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/host/bin/ace_gperf -DGEN_IDL_MAKE_DEPS -Wb,versioning_begin=TAO_BEGIN_VERSIONED_NAMESPACE_DECL -Wb,versioning_end=TAO_END_VERSIONED_NAMESPACE_DECL -Gp -Gd -Sci -SS -Sorb -GA -Wb,export_macro=TAO_PortableServer_Export -Wb,export_include=tao/PortableServer/portableserver_export.h -iC tao/PortableServer ForwardRequest.pidl
ForwardRequest.pidl:16:1: error: 'module' does not name a type
 module PortableServer
 ^~~~~~
/projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/host/bin/tao_idl: preprocessor "powerpc-604-linux-gnu-g++" returned with an error
GNUmakefile.PortableServer_Idl:235: recipe for target 'ForwardRequestA.cpp' failed
make[2]: *** [ForwardRequestA.cpp] Error 1
make[2]: Leaving directory '/projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/target/TAO/tao/PortableServer'
GNUmakefile.dist:208: recipe for target 'PortableServer_Idl' failed
make[1]: *** [PortableServer_Idl] Error 2
make[1]: Leaving directory '/projects/VSPx/tools/dds/OpenDDS/OpenDDS-3.15_ppc-604-linux/build/target'
GNUmakefile:3: recipe for target 'all' failed
make: *** [all] Error 2



Regards,

Paul
[OpenDDS-3.15_ppc-604-linux_build-log-2021-01-25_04-12-10.zip](https://github.com/objectcomputing/OpenDDS/files/5876341/OpenDDS-3.15_ppc-604-linux_build-log-2021-01-25_04-12-10.zip)

**Comment by mitza-oci** [comment]
This particular error looks like the preprocessing step of tao_idl is not working.  tao_idl will spawn a child process to run the C++ preprocessor on the IDL input file.  tao_idl uses `-E` or the equivalent for the specific system to tell the compiler to run the preprocessing step only.

Try removing `--macros=TAO_IDL_PREPROCESSOR=powerpc-604-linux-gnu-g++ "--macros=TAO_IDL_PREPROCESSOR_ARGS=-std=c++11 -I /projects/VSPx/tools/uei/ueipac-4.0.11/include"` from your configuration.

**Comment by PaulA-V** [comment]
I removed the macros from my configuration and the same error occurred (see attached logs).

GNUmakefile: /projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/target/TAO/tao/PortableServer/GNUmakefile.PortableServer_Idl MAKEFLAGS=w

/projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/host/bin/tao_idl -Wb,pre_include=ace/pre.h -Wb,post_include=ace/post.h -I../.. -g /projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/host/bin/ace_gperf -DGEN_IDL_MAKE_DEPS -Wb,versioning_begin=TAO_BEGIN_VERSIONED_NAMESPACE_DECL -Wb,versioning_end=TAO_END_VERSIONED_NAMESPACE_DECL -Gp -Gd -Sci -SS -Sorb -GA -Wb,export_macro=TAO_PortableServer_Export -Wb,export_include=tao/PortableServer/portableserver_export.h -iC tao/PortableServer ForwardRequest.pidl
ForwardRequest.pidl:16:1: error: ‘module’ does not name a type
 module PortableServer
 ^
/projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/host/bin/tao_idl: preprocessor "g++" returned with an error
GNUmakefile.PortableServer_Idl:235: recipe for target 'ForwardRequestA.cpp' failed
make[2]: *** [ForwardRequestA.cpp] Error 1
make[2]: Leaving directory '/projects/VSPx/tools/comms/ace+tao/ACE+TAO-2.2a_with_latest_patches_ppc-604-linux/ACE_wrappers/build/target/TAO/tao/PortableServer'
GNUmakefile.dist:208: recipe for target 'PortableServer_Idl' failed
make[1]: *** [PortableServer_Idl] Error 2
make[1]: Leaving directory '/projects/VSPx/tools/dds/OpenDDS/OpenDDS-3.15_ppc-604-linux/build/target'
GNUmakefile:3: recipe for target 'all' failed
make: *** [all] Error 2


Thank you for your help thus far.

Regards,

Paul
[OpenDDS-3.15_ppc-604-linux_build-log-2021-01-26_09-17-54.zip](https://github.com/objectcomputing/OpenDDS/files/5877382/OpenDDS-3.15_ppc-604-linux_build-log-2021-01-26_09-17-54.zip)

**Comment by mitza-oci** [comment]
Try running the same `tao_idl` command with an additional `-v` argument.  This will cause it to log the preprocessor command line.

**Comment by PaulA-V** [comment]
After rerunning the compile with the -v argument to tao_idl I found that the preprocessor arguments that I specified using the TAO_IDL_PREPROCESSOR_ARGS was present in the preprocessor call (-std=c++11 -I UEIPACROOT).  It appears I am a victim of self inflicted wounds.  I had forgotten that I set this variable in the shell environment under which the build was run.  So with this variable and the ARCH variable unset I performed a clean build which completed without error (see attached logs).

Thanks again Adam for your help on this.  Now it's on to testing OpenDDS on the target system!


Regards,

Paul
[OpenDDS-3.15_ppc-604-linux_build-log-2021-01-27_06-00-19.zip](https://github.com/objectcomputing/OpenDDS/files/5884512/OpenDDS-3.15_ppc-604-linux_build-log-2021-01-27_06-00-19.zip)
